pkg/client/unversioned/clientcmd/client_config.go

```diff
@@ -107,7 +110,11 @@ func (config *DirectClientConfig) ClientConfig() (*restclient.Config, error) {
 		// mergo is a first write wins for map value and a last writing wins for interface values
 		// NOTE: This behavior changed with https://github.com/imdario/mergo/commit/d304790b2ed594794496464fadd89d2bb266600a.
 		//       Our mergo.Merge version is older than this change.
-		userAuthPartialConfig, err := getUserIdentificationPartialConfig(configAuthInfo, config.fallbackReader)
+		var persister restclient.AuthProviderConfigPersister
+		if config.persistAuthProviderConfig != nil {
+			persister = config.persistAuthProviderConfig(config.getAuthInfoName())
+		}
+		userAuthPartialConfig, err := getUserIdentificationPartialConfig(configAuthInfo, config.fallbackReader, persister)
```

Comment:
> @cjcullen: I passed it through the DirectClientConfig so that I could plumb through the loadingRules from the DeferredLoadingClientConfig in merged_client_builder.go. I assumed this was necessary to make sure the old config gets loaded correctly. Was that a bad assumption?

I think the API will be easier to use if the `ClientConfig` has a method to get its `ConfigAccess`, which I'm pretty sure has enough enough information to modify the config.  Looks like you've already got one in the persisting struct, but I think its locations aren't being set to match to loading rules.